Update: Accept array for `ignorePattern` (fixes #3982)

jmm opened the pull request:
* Add test for `ignorePattern` array.
* Accept array for `ignorePattern`.
* Document `ignorePattern`.

ilyavolodin commented:
LGTM. Although I can swear I've seen this change before somehow.

nzakas commented on tests/lib/ignored-paths.js:
```diff
@@ -45,6 +45,17 @@ describe("IgnoredPaths", function() {
+                ignorePattern: ignorePattern
+            });
+            assert.ok(ignorePattern.every(function(patt) {
+                return ignoredPaths.patterns.indexOf(patt) >= 0;
```
Should this be `path` instead of `patt`?

jmm replied:
It's not a typo if that's what you mean. I commonly abbreviate "pattern" as "patt" (each of these being a glob pattern). But if you find it unclear it can be `pattern|glob` or whatever you prefer.

nzakas replied:
Yeah, I'd prefer either of those. Clarity > brevity. :)

jmm replied:
No problem, I pushed a new commit with `pattern`.

gyandeeps commented:
Dont we want to make sure we can also pass an array using CLI also?
Right now you can only pass in a string: https://github.com/eslint/eslint/blob/master/lib/options.js#L109

I think to be consistent, we should make this change for CLI also, thoughts @eslint/eslint-team ?

nzakas commented:
Probably a good idea.

nzakas commented:
@jmm can you make this change for the CLI as well?

jmm commented:
@nzakas I'll check it out when I get a chance.

nzakas commented:
@jmm do you have an ETA?

jmm commented:
@nzakas no, I don't know when I'll get a chance.

nzakas commented:
OK, then we will likely need to take this over for you. Thanks!

jmm commented:
@nzakas I'm actually squeezing this in now -- I expect to have it today.

jmm commented:
Alright, this should be all set.

It's too bad there's not a better way to test it via the CLI than checking if `console.log` was called.

Is there a better solution than duplicating the `syntax-error.js` fixture to `syntax-error2.js` like I did?

I don't know a legit way to test the rendering of the docs I added.

nzakas commented on tests/lib/cli.js:
```diff
@@ -358,6 +358,28 @@ describe("cli", function() {
     });
 
 
+    describe("when given patterns to ignore", function() {
```
See https://github.com/eslint/eslint/blob/master/tests/lib/cli.js#L647 for a better way to test (note https://github.com/eslint/eslint/blob/master/tests/lib/cli.js#L636)

jmm commented:
Thanks. I hadn't seen that note (I just went to the most similar existing test and started from there). What do you think of this (44a3662)? I originally set it up by spying on `CLIEngine.prototype.executeOnFiles()`, but that's really an implementation detail that I'd prefer not to base the test on. So I changed it to configure a custom formatter and spy on that. It's pretty hokey, but at least it's actually relying on public API.

A couple of other things:

* I borrowed an existing, unrelated fixture file. (I just needed an additional file to ignore).

* Do you know that sinon `sandbox.verifyAndRestore` isn't documented? https://github.com/eslint/eslint/blob/4a115f5cef5cb3605a3f650519d564b6614f8232/tests/lib/cli.js#L644 

I don't know what's up with the CI failure. I don't think it's related to anything I did. I didn't even touch the file with the failing test (which BTW it'd be great if it would report in the output).

IanVS commented:
Restarted Travis and the tests are now passing.

nzakas commented:
@jmm that seems like overkill. All you need to do is make sure the correct options are passed to CLIEngine.

jmm commented:
@nzakas Ok, check out 06eefb8.

nzakas commented:
We already have a pattern for testing that the correct arguments are passed (the test I referenced in https://github.com/eslint/eslint/blob/master/tests/lib/cli.js#L647). I'd like to keep this pattern consistent so it's easier for others to find multiple examples.

jmm commented:
@nzakas May I ask what is the purpose (other than consistency with other tests that also don't appear to benefit from it) of doing it in a way that requires stubbing a bunch of methods just to test that the correct arguments are passed to the constructor?

nzakas commented:
Yeah, we're trying to create a unit test instead of an integration test. We want to test the interfaces between the components are behaving correctly.

jmm commented:
@nzakas Is this accurate?:

> All you need to do is make sure the correct options are passed to CLIEngine.

Or do I also need to make sure it goes through the rest of the motions with data that replicates what should actually be flowing through?

nzakas commented:
Yes, that's accurate. Once `cli` passes options to `CLIEngine`, the rest is on `CLIEngine` to do the right thing.

jmm commented:
@nzakas Please see if ce88842 is what you're looking for. If so, do you at least want to extract a bunch of that boilerplate to a function that can be reused for all of the equivalent tests? To me it's really undesirable to have a bunch of duplicated code (and corresponding brittleness) by inlining the same boilerplate in a bunch of tests that are just to verify that the correct options hash arrives at CLIEngine.

nzakas commented:
Looks good. I'm open to consolidating if you want to take a stab at it.

jmm commented:
@nzakas This is the kind of thing I have in mind: https://github.com/jmm/eslint/commit/dc05a6478811d4f33a137f6e06468e76e0da511d.

Since the goal is just to verify what's being received by CLIEngine, it doesn't make sense to assert the return value of `cli.execute()`, right?

You intentionally want even single `it()`s wrapped in `describe()`, right?

BTW, is there supposed to be a max line length in this repo?

nzakas commented:
That looks like a nice start. 

> Since the goal is just to verify what's being received by CLIEngine, it doesn't make sense to assert the return value of cli.execute(), right?

Correct, there's no point in testing exit code. 

> You intentionally want even single it()s wrapped in describe(), right?

Yes

> BTW, is there supposed to be a max line length in this repo?

No.  I always just tend to break when I get annoyed. :)

jmm commented:
@nzakas Do you want me to squash that in here?

> No. I always just tend to break when I get annoyed. :)

Ok, thanks.

nzakas commented:
Yeah, feel free to add to this PR. I'm sure it will be a big help

jmm commented:
@nzakas Ok, thanks, please check out the latest. (I also updated the note comment about adding new tests and moved it to the top [though realistically I would still expect it to be overlooked much of the time]).

nzakas commented on tests/lib/cli.js:
```diff
@@ -357,6 +388,21 @@ describe("cli", function() {
+
+            verifyCLIEngineOpts(
+                ignorePaths.map(function(ignorePath) {
+                    return "--ignore-pattern " + ignorePath;
```
Can you move this out to a local variable? It will make it easier to see what you're doing here.

jmm commented:
@nzakas Is [this](https://github.com/jmm/eslint/blob/7446e4722e52847c874dd3c9086470c625f924fe/tests/lib/cli.js#L401-L409) what you're looking for?

jmm commented:
Thanks!
